contrib/gocql/gocql/observer.go

```diff
@@ -0,0 +1,123 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+package gocql
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/gocql/gocql"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+// NewTracedSession returns a new session augmented with tracing.
+func NewTracedSession(cluster *gocql.ClusterConfig, opts ...WrapOption) (*gocql.Session, error) {
+	obs := NewObserver(cluster, opts...)
+	cfg := obs.cfg
+
+	if cfg.traceQuery {
+		cluster.QueryObserver = obs
+	}
+	if cfg.traceBatch {
+		cluster.BatchObserver = obs
+	}
+	if cfg.traceConnect {
+		cluster.ConnectObserver = obs
+	}
+	return cluster.CreateSession()
+}
+
+// TraceQuery can be used to enable tracing an individual *gocql.Query.
+func TraceQuery(q *gocql.Query, cluster *gocql.ClusterConfig, opts ...WrapOption) *gocql.Query {
```

Comment:
^ I am not sure about this one. I know I may have been the one suggesting such a shortcut, but if it requires to keep track of all `*gocql.ClusterConfig` and `...WrapOption`, it's not going to be *THAT* convenient.

On top of this, creating an observer on-the-fly may just create side-activity (garbage collection because many objects created etc.) that is not optimal. In practice you'd want to create an observer for your session, and re-use it afterwards. It may not be harder to keep track of it than to keep track of both the config and the options.

At the end of the day, I'd shoot for providing users a "best practice" such as:

* if you want to not-trace-by-default then create a regular session, don't worry about it, and then just create your observer upfront while creating the session, eg:

```go
session,err := gocql.CreateSession(cfg)
// handle err
obs := gocqltrace.NewObserver(cfg,  gocqltrace.WithServiceName("my-cassandra-cluster"))
query := session.Query("SELECT something FROM somewhere").Observer(obs) // this is traced
query := session.Query("SELECT something FROM somewhere") // this is not traced
```
And if you want to trace by default
```go
session,err := gocqltrace.NewTracedSession(cfg, gocqltrace.WithServiceName("my-cassandra-cluster"),  gocqltrace.WithTraceQuery(true), gocqltrace.WithTraceBatch(true), gocqltrace.WithTraceConnect(true))
// handle err
query := session.Query("SELECT something FROM somewhere") // this is traced
query := session.Query("SELECT something FROM somewhere").Observer(nil) // this is not traced
```
I don't think it's more code than the current `TraceBatch` + `TraceQuery` option, and it's more versatile/symmetric, allowing users to just enable/disable tracing at will. Which was not possible before, but can be done quite easily now with this observer pattern.

I think it is fine to use `.Observer(nil)` when I read [this](https://github.com/apache/cassandra-gocql-driver/blob/974fa1211ccef10a418966fd7b00984a39ed9232/session.go#L957) [and this](https://github.com/apache/cassandra-gocql-driver/blob/974fa1211ccef10a418966fd7b00984a39ed9232/session.go#L1045)

Reply:
agree! this change makes sense and it simplifies stuff 👍